# Top 500 Movies Ranked by Critics and Audience scores
This notebook documents the cleaning and preparation of the Kaggle dataset. The aim is to produce a consistent, accurate, and analysis‑ready datast by resolving missing values, duplicates and formatting issues. This ensures reliable insights into movie rankings and critic vs. audience trends.

## 1. Data Inspection and Initial Assessment
We begin by importing essential Python libraries for data cleaning and analysis

In [1]:
import pandas as pd 
import numpy as np 

We load the raw employee dataset into a pandas DataFrame for inspection.

In [3]:
df = pd.read_csv("Top 500 Movies Ranked by Combined Critics and Audience Scores.csv")

### Initial Data Inspection

After loading the dataset, we perform a quick check to understand its structure, data types, missing values, and duplicate records.

In [4]:
df.head()

,Rank,Title,Year,Genre,Director,Cast,Language,Plot,Awards,Production,...,Critic_Reviews,Audience_Rating,Audience_Reviews,Letterboxd,Letterboxd_Votes,Google_Score,Streaming_On,RT_URL,imdbID,Custom_Score
0,1,The Godfather,1972,"Drama, Crime",Francis Ford Coppola,"Marlon Brando, Al Pacino, James Caan, Robert D...",en,"Spanning the years 1945 to 1955, a chronicle o...",Won 3 Oscars. Another 30 wins & 19 nominations.,Paramount Pictures,...,104,95.0,1087289.0,90,1789555.0,90.0,"paramount, itunes, amzvid",https://www.rottentomatoes.com/m/godfather,tt0068646,93.83
1,2,Seven Samurai,1954,"Action, Drama",Akira Kurosawa,"Toshirō Mifune, Takashi Shimura, Yoshio Inaba,...",ja,A samurai answers a village's request for prot...,Nominated for 2 Oscars. Another 6 wins & 5 nom...,Columbia Pictures,...,68,94.0,167123.0,92,318537.0,93.0,"hbo, itunes, amzvid, hbm",https://www.rottentomatoes.com/m/seven_samurai...,tt0047478,93.33
2,3,12 Angry Men,1957,Drama,Sidney Lumet,"Martin Balsam, John Fiedler, Lee J. Cobb, E.G....",en,The defense and the prosecution have rested an...,Nominated for 3 Oscars. Another 16 wins & 8 no...,Criterion Collection,...,58,94.0,376697.0,92,917590.0,91.0,"itunes, amzvid",https://www.rottentomatoes.com/m/1000013_12_an...,tt0050083,92.83
3,4,Schindler's List,1993,"Drama, History, War",Steven Spielberg,"Liam Neeson, Ben Kingsley, Ralph Fiennes, Caro...",en,The true story of how businessman Oskar Schind...,Won 7 Oscars. Another 74 wins & 26 nominations.,Universal Pictures,...,82,94.0,601555.0,90,978054.0,93.0,"itunes, amzvid",https://www.rottentomatoes.com/m/schindlers_list,tt0108052,92.50
4,5,The Godfather: Part II,1974,"Drama, Crime",Francis Ford Coppola,"Al Pacino, Robert Duvall, Diane Keaton, Robert...",en,In the continuing saga of the Corleone crime f...,Won 6 Oscars. Another 13 wins & 16 nominations.,Paramount Pictures,...,80,94.0,586600.0,92,934120.0,89.0,"paramount, itunes, amzvid",https://www.rottentomatoes.com/m/godfather_par...,tt0071562,91.67


In [5]:
df.tail()


,Rank,Title,Year,Genre,Director,Cast,Language,Plot,Awards,Production,...,Critic_Reviews,Audience_Rating,Audience_Reviews,Letterboxd,Letterboxd_Votes,Google_Score,Streaming_On,RT_URL,imdbID,Custom_Score
444,445,Rocky,1976,Drama,John G. Avildsen,"Sylvester Stallone, Talia Shire, Burt Young, C...",en,An uneducated collector for a Philadelphia loa...,Won 3 Oscars. Another 17 wins & 20 nominations.,United Artists,...,59,71.0,22486658.0,82,651986.0,90.0,"itunes, amzvid",https://www.rottentomatoes.com/m/1017776-rocky,tt0075148,79.17
445,446,Phantom Thread,2017,"Drama, Romance",Paul Thomas Anderson,"Daniel Day-Lewis, Vicky Krieps, Lesley Manvill...",en,"In 1950s London, a renowned dressmaker's metic...",NaN,NaN,...,235,71.0,158555.0,82,459029.0,72.0,"itunes, amzvid",https://www.rottentomatoes.com/m/phantom_thread,tt5776858,79.17
446,447,Paterson,2016,"Drama, Comedy, Romance",Jim Jarmusch,"Adam Driver, Golshifteh Farahani, Nellie, Rizw...",en,"A week in the life of Paterson, a poet bus dri...",NaN,NaN,...,177,71.0,74548.0,78,190980.0,77.0,prime,https://www.rottentomatoes.com/m/paterson,tt5247022,79.17
447,448,Titanic,1997,"Drama, Romance",James Cameron,"Leonardo DiCaprio, Kate Winslet, Billy Zane, K...",en,101-year-old Rose DeWitt Bukater tells the sto...,Won 11 Oscars. Another 106 wins & 73 nominations.,Paramount Pictures,...,147,71.0,24549197.0,76,2052322.0,93.0,"paramount, itunes, amzvid",https://www.rottentomatoes.com/m/titanic/,tt0120338,79.00
448,449,Uncut Gems,2019,"Drama, Thriller, Crime","Josh Safdie, Benny Safdie","Adam Sandler, LaKeith Stanfield, Julia Fox, Ke...",en,Howard is a successful New York City jeweler f...,NaN,NaN,...,204,57.0,438510.0,78,1265981.0,74.0,"amzvid, itunes",https://www.rottentomatoes.com/m/uncut_gems,tt5727208,77.00


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 449 entries, 0 to 448
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Rank               449 non-null    int64  
 1   Title              449 non-null    object 
 2   Year               449 non-null    int64  
 3   Genre              449 non-null    object 
 4   Director           449 non-null    object 
 5   Cast               448 non-null    object 
 6   Language           449 non-null    object 
 7   Plot               449 non-null    object 
 8   Awards             324 non-null    object 
 9   Production         329 non-null    object 
 10  Flickmetrix_Score  449 non-null    float64
 11  IMDb_10            449 non-null    float64
 12  IMDb_100           449 non-null    float64
 13  IMDb_Votes         449 non-null    int64  
 14  Metacritic         436 non-null    float64
 15  Critic_Rating_RT   449 non-null    float64
 16  Critic_Reviews     449 non

In [7]:
df.columns

Index(['Rank', 'Title', 'Year', 'Genre', 'Director', 'Cast', 'Language',
       'Plot', 'Awards', 'Production', 'Flickmetrix_Score', 'IMDb_10',
       'IMDb_100', 'IMDb_Votes', 'Metacritic', 'Critic_Rating_RT',
       'Critic_Reviews', 'Audience_Rating', 'Audience_Reviews', 'Letterboxd',
       'Letterboxd_Votes', 'Google_Score', 'Streaming_On', 'RT_URL', 'imdbID',
       'Custom_Score'],
      dtype='object')

In [8]:
df.describe()

,Rank,Year,Flickmetrix_Score,IMDb_10,IMDb_100,IMDb_Votes,Metacritic,Critic_Rating_RT,Critic_Reviews,Audience_Rating,Audience_Reviews,Letterboxd,Letterboxd_Votes,Google_Score,Custom_Score
count,449.000000,449.000000,449.000000,449.000000,449.000000,4.490000e+02,436.000000,449.000000,449.000000,401.000000,4.420000e+02,449.000000,4.480000e+02,446.000000,449.000000
mean,225.000000,1994.151448,83.884187,7.969488,79.694878,4.315229e+05,86.754587,85.581292,122.926503,85.748130,8.115906e+05,82.151448,7.029797e+05,85.807175,84.229866
std,129.759393,25.683153,2.685102,0.359911,3.599106,4.723575e+05,6.924011,4.844751,67.931731,5.018357,3.179307e+06,3.954457,7.575711e+05,5.459435,2.798203
min,1.000000,1920.000000,80.000000,7.100000,71.000000,1.194000e+04,65.000000,72.000000,50.000000,57.000000,1.759300e+04,68.000000,3.153500e+04,46.000000,77.000000
25%,113.000000,1976.000000,82.000000,7.700000,77.000000,1.019410e+05,82.000000,82.000000,70.000000,83.000000,7.716325e+04,80.000000,1.430270e+05,83.000000,82.170000
50%,225.000000,2003.000000,83.000000,8.000000,80.000000,2.312560e+05,87.000000,85.000000,100.000000,87.000000,1.982740e+05,82.000000,3.962565e+05,86.000000,83.670000
75%,337.000000,2015.000000,85.000000,8.200000,82.000000,6.460640e+05,92.000000,89.000000,156.000000,89.000000,5.818142e+05,84.000000,1.002294e+06,90.000000,86.000000
max,449.000000,2025.000000,93.000000,9.200000,92.000000,2.666368e+06,100.000000,98.000000,342.000000,95.000000,2.454920e+07,92.000000,3.771374e+06,98.000000,93.830000


In [9]:
df.isnull().sum()

Rank                   0
Title                  0
Year                   0
Genre                  0
Director               0
Cast                   1
Language               0
Plot                   0
Awards               125
Production           120
Flickmetrix_Score      0
IMDb_10                0
IMDb_100               0
IMDb_Votes             0
Metacritic            13
Critic_Rating_RT       0
Critic_Reviews         0
Audience_Rating       48
Audience_Reviews       7
Letterboxd             0
Letterboxd_Votes       1
Google_Score           3
Streaming_On          30
RT_URL                17
imdbID                 0
Custom_Score           0
dtype: int64

In [10]:
df.duplicated().sum()

np.int64(0)

###  Observations During Data Cleaning

- **Column Names**
  - Need to be stripped and converted to lowercase.
  - Underscores are already present and consistent, so no change needed.

- **Irrelevant Columns**
  - Dropped: `plot`, `awards`, `RT_URL`, `IMDb_10`.
  - Rationale: `IMDb_10` and `IMDb_100` represent the same rating on different scales.  
    Since other scores are out of 100, `IMDb_100` is retained for consistency.

- **Missing Values**
  - `awards` → 125 missing  
  - `production` → 120 missing  
  - `cast` → 1 missing  
  - `metacritic` → 13 missing  
  - `audience_rating` → 48 missing  
  - `audience_reviews` → 7 missing  
  - `letterboxd_votes` → 1 missing  
  - `google_score` → 3 missing  
  - `streaming_on` → 30 missing  
  - `RT_URL` → 17 missing  

- **Duplicates**
  - No duplicate records found.

- **Data Type Conversion**
  - `audience_reviews` → float → int  
  - `letterboxd_votes` → float → int  

- **Text Standardization**
  - Normalize spelling, capitalization, and formatting across text fields.
  - Standardize `language` codes (e.g., `en` → English, `ja` → Japanese).

- **Quality Checks (before final verification)**
  - **Consistency checks** → Ensure related fields align (e.g., critic vs audience scores).  
  - **Range validation** → Confirm numeric values fall within valid ranges (scores 0–100, years realistic).  

- **Final Validation**
  - Re-run `.info()` and `.describe()` to confirm dataset consistency after cleaning.

## 2. Clean and Standardize Column Names
To ensure consistency across the dataset, all column names will be converted to lowercase and stripped of leading/trailing spaces. This makes the DataFrame easier to work with and avoids errors during analysis.

In [12]:
df.columns= df.columns.str.strip().str.lower()
df.columns

Index(['rank', 'title', 'year', 'genre', 'director', 'cast', 'language',
       'plot', 'awards', 'production', 'flickmetrix_score', 'imdb_10',
       'imdb_100', 'imdb_votes', 'metacritic', 'critic_rating_rt',
       'critic_reviews', 'audience_rating', 'audience_reviews', 'letterboxd',
       'letterboxd_votes', 'google_score', 'streaming_on', 'rt_url', 'imdbid',
       'custom_score'],
      dtype='object')

## 3. Clean Whitespace in Text Columns

To ensure consistency, all text (string/object) columns will be stripped of leading and trailing whitespace.  
This prevents issues during analysis, such as mismatches in categorical values caused by hidden spaces.

In [16]:
for col in df.select_dtypes(include='object').columns:
    df[col]=df[col].astype(str).str.strip().str.lower()
df.head()



,rank,title,year,genre,director,cast,language,plot,awards,production,...,critic_reviews,audience_rating,audience_reviews,letterboxd,letterboxd_votes,google_score,streaming_on,rt_url,imdbid,custom_score
0,1,the godfather,1972,"drama, crime",francis ford coppola,"marlon brando, al pacino, james caan, robert d...",en,"spanning the years 1945 to 1955, a chronicle o...",won 3 oscars. another 30 wins & 19 nominations.,paramount pictures,...,104,95.0,1087289.0,90,1789555.0,90.0,"paramount, itunes, amzvid",https://www.rottentomatoes.com/m/godfather,tt0068646,93.83
1,2,seven samurai,1954,"action, drama",akira kurosawa,"toshirō mifune, takashi shimura, yoshio inaba,...",ja,a samurai answers a village's request for prot...,nominated for 2 oscars. another 6 wins & 5 nom...,columbia pictures,...,68,94.0,167123.0,92,318537.0,93.0,"hbo, itunes, amzvid, hbm",https://www.rottentomatoes.com/m/seven_samurai...,tt0047478,93.33
2,3,12 angry men,1957,drama,sidney lumet,"martin balsam, john fiedler, lee j. cobb, e.g....",en,the defense and the prosecution have rested an...,nominated for 3 oscars. another 16 wins & 8 no...,criterion collection,...,58,94.0,376697.0,92,917590.0,91.0,"itunes, amzvid",https://www.rottentomatoes.com/m/1000013_12_an...,tt0050083,92.83
3,4,schindler's list,1993,"drama, history, war",steven spielberg,"liam neeson, ben kingsley, ralph fiennes, caro...",en,the true story of how businessman oskar schind...,won 7 oscars. another 74 wins & 26 nominations.,universal pictures,...,82,94.0,601555.0,90,978054.0,93.0,"itunes, amzvid",https://www.rottentomatoes.com/m/schindlers_list,tt0108052,92.50
4,5,the godfather: part ii,1974,"drama, crime",francis ford coppola,"al pacino, robert duvall, diane keaton, robert...",en,in the continuing saga of the corleone crime f...,won 6 oscars. another 13 wins & 16 nominations.,paramount pictures,...,80,94.0,586600.0,92,934120.0,89.0,"paramount, itunes, amzvid",https://www.rottentomatoes.com/m/godfather_par...,tt0071562,91.67


## 4. Text Standardization

To ensure categorical values are consistent, text fields will be standardized by:
- Converting all text to lowercase *(already completed)*.
- Mapping language codes (e.g., `en` → English, `ja` → Japanese).
- Removing extra spaces or punctuation inconsistencies *(done again after mapping to catch any new issues).*

This step ensures uniformity across columns such as `genre`, `director`, `cast`, and `language`.

In [27]:
'''Standardize Language codes'''

unique_code = df["language"].unique()

In [21]:
language_map = {"en" : "english" ,"ja": "japanese", "ko": "korean", "it": "italian",
    "fr": "french", "fa": "persian", "de": "german", "es": "spanish",
    "sv": "swedish", "pt": "portuguese", "cn": "chinese", "ru": "russian",
    "id": "indonesian", "zh": "chinese", "lv": "latvian", "he": "hebrew",
    "ro": "romanian", "ga": "irish", "te": "telugu", "da": "danish",
    "ar": "arabic", "tr": "turkish", "no": "norwegian", "pl": "polish",
    "hi": "hindi", "gl": "galician", "hu": "hungarian"}
unmapped_codes = [code for code in unique_code if code not in language_map]
print("Unmapped codes:\n" , unmapped_codes)



Unmapped codes:
 []


In [25]:
df['language'] = df['language'].replace(language_map)

In [26]:
df['language'].unique()

array(['english', 'japanese', 'korean', 'italian', 'french', 'persian',
       'german', 'spanish', 'swedish', 'portuguese', 'chinese', 'russian',
       'indonesian', 'latvian', 'hebrew', 'romanian', 'irish', 'telugu',
       'danish', 'arabic', 'turkish', 'norwegian', 'polish', 'hindi',
       'galician', 'hungarian'], dtype=object)

In [29]:
''' Remove multiple spaces inside text'''
for col in df.select_dtypes(include='object').columns:
    df[col]=df[col].str.replace(r"\s+"," ",regex=True)

In [32]:
'''Remove unwanted punctuation'''

import string

punctuation_to_remove = string.punctuation

for col in df.select_dtypes(include="object").columns:
    df[col]=df[col].str.replace(f"[{punctuation_to_remove}]","", regex=True)

In [33]:
'''Remove whitespace again'''
for col in df.select_dtypes(include="object").columns:
    df[col]=df[col].str.strip()

## 5. Dropping Irrelevant Columns

Reminged irrelevant columns (`plot`, `awards`, `RT_URL`, `IMDb_10`) to retain consistency and make analysis cleaner and more focused.

In [46]:
irrelevant_columns = ["streaming_on", "plot","awards","imdb_10","cast","director","production","rt_url"]
df = df.drop(columns=irrelevant_columns,errors="ignore")


In [45]:
df.columns

Index(['rank', 'title', 'year', 'genre', 'language', 'flickmetrix_score',
       'imdb_100', 'imdb_votes', 'metacritic', 'critic_rating_rt',
       'critic_reviews', 'audience_rating', 'audience_reviews', 'letterboxd',
       'letterboxd_votes', 'google_score', 'imdbid', 'custom_score'],
      dtype='object')

## 6. Standardizing Missing Values

Converted inconsistent placeholders (like "NA", "null", "n/a") into `NaN` 
to ensure consistent handling of missing data across all columns.

In [49]:
na_values = [ "na","nan"," ", "null"]
for value in df.select_dtypes(include="object").columns:
    df[value]=df[value].replace(na_values,np.nan)

## 7. Handling Missing Values
 
Numeric columns with missing values were imputed using the median to preserve distribution.  
Categorical/text columns were filled with `"Unknown"` to retain completeness without bias.
.

In [52]:
'''Replacing missing values with Median'''
numerical_col=["metacritic", "audience_rating", 
                "audience_reviews", "letterboxd_votes", 
                "google_score"]
for col in numerical_col:
    df[col]=df[col].fillna(df[col].median())


In [54]:
'''Replacing text missing values with unknown'''
for col in df.select_dtypes(include="object").columns:
    df[col]=df[col].fillna("unknown")
   

In [55]:
df.isnull().sum()

rank                 0
title                0
year                 0
genre                0
language             0
flickmetrix_score    0
imdb_100             0
imdb_votes           0
metacritic           0
critic_rating_rt     0
critic_reviews       0
audience_rating      0
audience_reviews     0
letterboxd           0
letterboxd_votes     0
google_score         0
imdbid               0
custom_score         0
dtype: int64

## 8. Data Type Conversion

Converted `audience_reviews` and `letterboxd_votes` from float to intege 
to ensure consistency in numeric representation and simplify analysis.


In [56]:
df['audience_reviews']=df['audience_reviews'].astype("int64")
df['letterboxd_votes']=df['letterboxd_votes'].astype("int64")

In [58]:
df.dtypes

rank                   int64
title                 object
year                   int64
genre                 object
language              object
flickmetrix_score    float64
imdb_100             float64
imdb_votes             int64
metacritic           float64
critic_rating_rt     float64
critic_reviews         int64
audience_rating      float64
audience_reviews       int64
letterboxd             int64
letterboxd_votes       int64
google_score         float64
imdbid                object
custom_score         float64
dtype: object

## 9. Quality Checks (before final verification)

Performed consistency checks to ensure related fields align (e.g., critic vs audience scores) and applied range validation to confirm numeric values fall within valid ranges (scores 0–100, years realistic).

In [81]:
'''Consistency check: critic vs audience rating should not widely differ'''
score_diff = abs(df["critic_rating_rt"]-df["audience_rating"])
print(df[score_diff>50][["title","critic_rating_rt","audience_rating"]])


Empty DataFrame
Columns: [title, critic_rating_rt, audience_rating]
Index: []


In [83]:
''' Range validation: ensure scores are between 0 and 100 '''
for col in ["flickmetrix_score","imdb_100","metacritic","critic_rating_rt","audience_rating","google_score"]:
    invalid = df[(df[col]<0) | (df[col]>100)]
    if not invalid.empty:
        print(f"Invalid values in {col}:\n", invalid[[col]])


In [88]:
''' Year validation'''
invalid_year= df[(df["year"]<1900) | (df["year"]>2026)]
print(invalid_year[["title","year"]])

Empty DataFrame
Columns: [title, year]
Index: []


## Final Data Quality Review

In [89]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 449 entries, 0 to 448
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   rank               449 non-null    int64  
 1   title              449 non-null    object 
 2   year               449 non-null    int64  
 3   genre              449 non-null    object 
 4   language           449 non-null    object 
 5   flickmetrix_score  449 non-null    float64
 6   imdb_100           449 non-null    float64
 7   imdb_votes         449 non-null    int64  
 8   metacritic         449 non-null    float64
 9   critic_rating_rt   449 non-null    float64
 10  critic_reviews     449 non-null    int64  
 11  audience_rating    449 non-null    float64
 12  audience_reviews   449 non-null    int64  
 13  letterboxd         449 non-null    int64  
 14  letterboxd_votes   449 non-null    int64  
 15  google_score       449 non-null    float64
 16  imdbid             449 non

In [90]:
df.columns

Index(['rank', 'title', 'year', 'genre', 'language', 'flickmetrix_score',
       'imdb_100', 'imdb_votes', 'metacritic', 'critic_rating_rt',
       'critic_reviews', 'audience_rating', 'audience_reviews', 'letterboxd',
       'letterboxd_votes', 'google_score', 'imdbid', 'custom_score',
       'score_diff'],
      dtype='object')

In [91]:
df.isnull().sum()

rank                 0
title                0
year                 0
genre                0
language             0
flickmetrix_score    0
imdb_100             0
imdb_votes           0
metacritic           0
critic_rating_rt     0
critic_reviews       0
audience_rating      0
audience_reviews     0
letterboxd           0
letterboxd_votes     0
google_score         0
imdbid               0
custom_score         0
score_diff           0
dtype: int64

In [92]:
df.duplicated().sum()

np.int64(0)

In [93]:
df.head()

,rank,title,year,genre,language,flickmetrix_score,imdb_100,imdb_votes,metacritic,critic_rating_rt,critic_reviews,audience_rating,audience_reviews,letterboxd,letterboxd_votes,google_score,imdbid,custom_score,score_diff
0,1,the godfather,1972,drama crime,english,93.0,92.0,2108796,100.0,96.0,104,95.0,1087289,90,1789555,90.0,tt0068646,93.83,1.0
1,2,seven samurai,1954,action drama,japanese,93.0,86.0,381712,98.0,97.0,68,94.0,167123,92,318537,93.0,tt0047478,93.33,3.0
2,3,12 angry men,1957,drama,english,92.0,90.0,917215,97.0,93.0,58,94.0,376697,92,917590,91.0,tt0050083,92.83,1.0
3,4,schindlers list,1993,drama history war,english,92.0,90.0,1512905,95.0,93.0,82,94.0,601555,90,978054,93.0,tt0108052,92.50,1.0
4,5,the godfather part ii,1974,drama crime,english,92.0,90.0,1420179,90.0,95.0,80,94.0,586600,92,934120,89.0,tt0071562,91.67,1.0


## Final Conclusion
The dataset is now:

- **Cleaned** → missing values handled, invalid entries corrected  
- **Standardized** → column names, data types, and formats aligned  
- **Structurally consistent** →  categories validated, logical ranges confirmed  
- **Ready for analysis** → reliable and trustworthy for further exploration, visualization, or modeling

In [94]:
df.to_csv(r"C:\Users\garim\OneDrive\Desktop\github projects\data cleaning with python\500 movies ranked by imdb and others\Cleaned_Movie_Dataset.csv", index=False)
